linalg/dual.hpp

```diff
@@ -0,0 +1,413 @@
+}
+
+/** @brief implementation of tan for dual numbers */
+template <typename value_type, typename gradient_type> MFEM_HOST_DEVICE
+dual<value_type, gradient_type> tan(dual<value_type, gradient_type> a)
+{
+   using std::tan;
+   value_type f = tan(a.value);
+   return {f, a.gradient * (value_type{1} + f * f)};
+}
+
+/** @brief implementation of atan for dual numbers */
+template <typename value_type, typename gradient_type> MFEM_HOST_DEVICE
+dual<value_type, gradient_type> atan(dual<value_type, gradient_type> a)
+{
+   return {atan(a.value), a.gradient / (value_type{1} + a.value * a.value)};
+}
+
+/** @brief implementation of exponential function for dual numbers */
+template <typename value_type, typename gradient_type> MFEM_HOST_DEVICE
+dual<value_type, gradient_type> exp(dual<value_type, gradient_type> a)
+{
+   return {std::exp(a.value), std::exp(a.value) * a.gradient};
+}
+
+/** @brief implementation of the natural logarithm function for dual numbers */
+template <typename value_type, typename gradient_type> MFEM_HOST_DEVICE
+dual<value_type, gradient_type> log(dual<value_type, gradient_type> a)
+{
+   return {std::log(a.value), a.gradient / a.value};
+}
+
+/** @brief implementation of `a` (dual) raised to the `b` (dual) power */
+template <typename value_type, typename gradient_type> MFEM_HOST_DEVICE
+dual<value_type, gradient_type> pow(dual<value_type, gradient_type> a,
+                                    dual<value_type, gradient_type> b)
+{
+   value_type value = pow(a.value, b.value);
+   return {value, value * (a.gradient * (b.value / a.value) + b.gradient * std::log(a.value))};
+}
+
+/** @brief implementation of `a` (non-dual) raised to the `b` (dual) power */
+template <typename value_type, typename gradient_type> MFEM_HOST_DEVICE
+dual<value_type, gradient_type> pow(double a, dual<value_type, gradient_type> b)
+{
+   value_type value = pow(a, b.value);
+   return {value, value * b.gradient * std::log(a)};
+}
+
+/** @brief implementation of `a` (non-dual) raised to the `b` (non-dual) power */
+template <typename value_type > MFEM_HOST_DEVICE
+value_type pow(value_type a, value_type b) { return std::pow(a, b); }
+
+/** @brief implementation of `a` (dual) raised to the `b` (non-dual) power */
+template <typename value_type, typename gradient_type> MFEM_HOST_DEVICE
+dual<value_type, gradient_type> pow(dual<value_type, gradient_type> a, double b)
+{
+   value_type value = pow(a.value, b);
+   return {value, value * a.gradient * b / a.value};
+}
+
+/** @brief overload of operator<< for `dual` to work with work with standard output streams */
+template <typename value_type, typename gradient_type, int... n>
+std::ostream& operator<<(std::ostream& os, dual<value_type, gradient_type> A)
+{
+   os << '(' << A.value << ' ' << A.gradient << ')';
+   return os;
+}
+
+/** @brief promote a value to a dual number of the appropriate type */
+MFEM_HOST_DEVICE constexpr dual<double, double> make_dual(double x) { return {x, 1.0}; }
```

Comment:
The gradient should be 0, not 1. This is the value consistent with the ctor from `value_type`.

Reply:
If the intent is to set the `gradient` to `1` that definitely needs to be documented. Otherwise, most people will assume that this function is the same as the `dual` constructor that takes a double value.